File: HW 07/src/Boards/BoardArray2D.java
package Boards;
/**
 * 
 * BoardArray2D class represents the board configuration with 2D Array
 * 
 * @author Berk Pekgoz GTU 171044041
 *
 */
public class BoardArray2D extends AbstractBoard{

	private int[][] board;

	/**
	 * No paramater constructor for BoardArray2D.
	 * Size of board is setted to (3,3) as default.
	 */
	public BoardArray2D(){
		this(3,3);
	}

	
	/**
	 * Constructor with size parameters.
	 * @param rowSize Parameter for row size
	 * @param columnSize Parameter for column size
	 */
	public BoardArray2D(int rowSize , int columnSize){
		super();
		board=null;
		if(rowSize < 3 || columnSize < 3){
			rowSize = 3 ; 
			columnSize = 3;
		}
		this.setSize(rowSize,columnSize);
		reset();
	}
	
	/**
	 * Copy Constructor with another BoardArray2D object referance
	 * @param other Board2D object referance for copy constructor
	 */
	public BoardArray2D(BoardArray2D other){
		increaseTotalBoard();
		setSize(other.getRowSize(),other.getColumnSize());

		for(int i=0; i < getRowSize(); i++){
			for(int j=0; j< getColumnSize(); j++){
				board[i][j] = other.board[i][j];
			}
		}
	}

	public void setSize(int row , int column){
		if(row < 3 || column < 3){
			row = 3 ; 
			column = 3;
		}
		super.setSize(row,column);
		board = new int[row][column];
		reset();
	}

	public int cell(int row , int column){
		if( !isCellValid(row,column))
			System.exit(1);
		return board[row][column];
	}

	protected void setCell(int row , int column , int data){
		if( !isCellValid(row,column))
			System.exit(1);
		board[row][column] = data;
	}



}

File: HW 05/BoardArray2D.cpp
#include "BoardArray2D.h"

namespace GTU_PKGZ{

	BoardArray2D::BoardArray2D():BoardArray2D(3,3){}
	//default constructer

	BoardArray2D::BoardArray2D(int rowInput, int columnInput):AbstractBoard(rowInput,columnInput),Board(nullptr){
		setSize(rowInput,columnInput);
		reset();
	}

	BoardArray2D::BoardArray2D(const BoardArray2D& other):AbstractBoard(other),Board(nullptr){
		increaseTotalBoard();
		this->setSize(getRowSize(),getColumnSize());
		for(int i=0; i < getRowSize(); i++){
			for(int j=0; j< getColumnSize(); j++){
				Board[i][j] = other.Board[i][j];
			}
		}
	}


	BoardArray2D::~BoardArray2D(){
		emptyBoard();
	}

	const BoardArray2D& BoardArray2D::operator=(const BoardArray2D& other){
		if(this == &other)
			return *this;

		emptyBoard();
		AbstractBoard::operator=(other);
		setSize(getRowSize(),getColumnSize());

		for(int i=0; i < getRowSize(); i++){
			for(int j=0; j< getColumnSize(); j++){
				Board[i][j] = other.Board[i][j];
			}
		}
		
		return *this;
	}

	void BoardArray2D::emptyBoard(){
			for(int i=0; i < getRowSize(); i++){
				delete [] Board[i];
			}
			delete [] Board;
			Board = nullptr;
	}

	void BoardArray2D::setSize(int rowInput,int columnInput){
		AbstractBoard::setSize(rowInput,columnInput);
		if(Board != nullptr){
			delete [] Board;
		}

		Board = new int*[rowInput];
		for(int i=0; i < rowInput; i++){
			Board[i] = new int[columnInput];
		}
		reset();
	}

	int& BoardArray2D::operator()(int row,int column){
		if( !isCellValid(row,column))
			exit(1);
		return Board[row][column];
	}

	const int& BoardArray2D::operator()(int row,int column)const{
		if( !isCellValid(row,column))
			exit(1);
		return Board[row][column];
	}
}